File: diretorioprincipal/questao4.java
//atividade 06/09/24
//aluno: João Pedro Araújo Inácio
//questao4

import java.util.ArrayList;
import java.util.List;

abstract class Animal {
    protected String nome;

    public Animal(String nome) {
        this.nome = nome;
    }

    public abstract String som(); 
}

class Cachorro extends Animal {
    public Cachorro(String nome) {
        super(nome);
    }

    @Override
    public String som() {
        return nome + " late ruf ruf";
    }
}

class Gato extends Animal {
    public Gato(String nome) {
        super(nome);
    }

    @Override
    public String som() {
        return nome + " mia meow";
    }
}

class questao4 {
    public static void emitirSons(List<Animal> animais) {
        for (Animal animal : animais) {
            System.out.println(animal.som());
        }
    }

    public static void main(String[] args) {
        Animal cachorro = new Cachorro("noah");
        Animal gato = new Gato("yoyo");

        List<Animal> animais = new ArrayList<>();
        animais.add(cachorro);
        animais.add(gato);

        emitirSons(animais);
    }
}

File: diretorioprincipal/questao4.py
#atividade 06/09/24
#aluno: João Pedro Araújo Inácio
#questao4

class Animal:
    def __init__(self, nome):
        self.nome = nome

class Cachorro(Animal):
    def som(self):
        return f"{self.nome} late ruf ruf"

class Gato(Animal):
    def som(self):
        return f"{self.nome} mia miau"


def emitir_sons(animais):
    for animal in animais:
        print(animal.som())

cachorro = Cachorro("noah")
gato = Gato("memel")

animais = [cachorro, gato]
emitir_sons(animais)